File: euler016-proc.php
#!/usr/bin/env php
<?php
/** Problem 16: Power digit sum
 * https://projecteuler.net/problem=16
 *
 * In a typical Project Euler fashion, this is a problem that sound simple
 * but the size of the input makes it impractical to bruteforce. PHP isn't
 * up to the task when it comes to computing 2^1000 directly, so this one
 * has to be approached as if done by hand on paper.
 */


// To make things readable, numbers have the least significant digit first
function doubleNumber(array $number): array
{
    $carry = 0;
    $result = [];
    for ($i = 0; $i < count($number); $i++) {
        $sum = $number[$i] * 2 + $carry;
        $digit = $sum % 10;
        $carry = ($sum - $digit) / 10;
        $result[] = $digit;
    }
    if ($carry > 0) {
        $result[] = $carry;
    }
    return $result;
}

function euler016(int $power): int
{
    $number = [2];
    for ($i = 1; $i < $power; $i++) {
        $number = doubleNumber($number);
    }
    return array_sum($number);
}

// Main
echo euler016(intval(trim(file_get_contents('php://STDIN')))) . PHP_EOL;

File: euler016-proc.py
#!/usr/bin/env python3

# Problem 16: Power digit sum
# https://projecteuler.net/problem=16

import sys

def double(number):
    result = []
    carry = 0
    for i in range(len(number)):
        total = number[i] * 2 + carry
        digit = total % 10
        carry = (total - digit) // 10
        result.append(digit)
    if carry > 0:
        result.append(carry)
    return result

def euler016(power):
    number = [2]
    for i in range(1, power):
        number = double(number)
    return sum(number)

def parse_input(lines):
    return int(lines[0].strip())

if __name__ == "__main__":
    print(euler016(parse_input(sys.stdin.readlines())))
